rfc: RFC 18 Guidance on Workflow for Research Code

nielsleadholm opened the pull request:
This RFC proposes some guidance on the workflow that TBP researchers (or community members doing research) can follow when contributing code to Monty.

nielsleadholm commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,102 @@
+- However, adhering to these guidelines is often not possible without first carrying out a significant refactor of the code, or through a time-consuming design process.
+- As such, researchers should not follow these guidelines if it will introduce significant delays. Rather, the guidance they provide (e.g., being mindful of how state is added or accessed in code) can provide general tips for writing high-quality code from the beginning.
+
+### Other Comments
```
Right. Re. protection, do you feel like this comment is insufficient? These forks are not like the main platform, where others are pulling from them and an accidental commit could be highly damaging. In the unlikely case that a commit ever happened by accident, it could always be reverted / the changes undone. 

Re. squash merging, that is mainly for keeping the history in `tbp.monty` concise, and any PRs that make their way into it will be squashed anyways, so again I'm not sure it's an issue? Feel free to say though if you're concerned.

nielsleadholm commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,102 @@
+    - You should formulate PRs into the main branch as _atomic_ PRs for review by other researchers.
+    - Note that _atomic_ does not always mean small, although it is a common finding that PRs (especially by researchers!) are too large. Rather, atomic means that the PR represents a full, minimally self-contained change. The PR should not depend on code that will be added in the future, although context can be provided where helpful re. future changes.
+- When you open a PR into your fork's `main` branch, tag at least one other TBP researcher for review.
+    - You should NOT merge PR changes into the Research Prototype `main` branch without receiving a review from another researcher.
```
My bias would be to try this way to begin with. Overall we are not putting in much structure into this workflow. Asking that PRs are atomic and are reviewed is pretty much the only structure, and I am concerned that without this, the process will become too uncontrolled. The undesirable default in research is often that someone works on a "PR" for ages, which turns into an enormous collection of code. When it's finally reviewed, it becomes clear that there was a misunderstanding or false direction that could have been avoided with earlier intervention / reviews. 

So as said my bias is to have this guidance for now. If we find it is slowing things down like you say, then we can revisit. Notably the guidance is for review by a researcher, not necessarily a manager. This was helpful during the DMC where Scott and Hojae would often review one-another's PRs related to the configs and visualization code. 

Curious for other's views including @vkakerbeck

vkakerbeck replied:
Maybe we could have a middle ground where potentially controversial changes (mainly things that conceptually change how Monty works) should be reviewed or at least discussed as early as possible. 
But I also agree that we don't want to slow research prototypes down by requiring a review on every atomic change (which usually takes more than 24 hours). 

It's hard to suggest a concrete guideline for this as I think it is something we need to weigh on a case-by-case basis. Maybe just something like "RP work should be regularly reviewed and discussed with other researchers. Researchers can open PRs from branches on the feature fork to the feature main branch to receive early feedback on the execution of their idea. The start of an implementation project (PR to tbp.monty) should not be the first time a new feature is reviewed by other researchers. However, the researcher should use their own judgement to weigh the benefits of reviews on every atomic change against the benefit of quick iteration and experimentation."

nielsleadholm commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,102 @@
+
+### Multiple Researchers Working on a Single Research Prototype
+- From time-to-time, two or more researchers may be working on ideas that could be considered elements of a single, over-arching Research Prototype. In this case, it is perfectly acceptable for them to use a shared RP fork.
+- As with typical Research Prototypes, this work should still regularly be synced with the upstream `tbp.monty`. The main difference is that the collaborating researchers can submit PRs into this shared fork.
```
Good point, I've added some text on this.

vkakerbeck commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,104 @@
+
+### Alternative Workflows
+- We discussed using branches within existing personal forks as a way of managing Research Prototypes. However, this risks becoming overwhelming with the number of features. It also does not provide the same visibility to the community about existing prototypes that are reaching a mature stage of development.
+
```
One other alternative that I really liked was to use the `monty_project_template` https://github.com/thousandbrainsproject/tbp.monty_project_template to start a project that just uses Monty as a dependency. I enjoyed that workflow for our ultrasound project as it clearly isolated all the changes made in custom classes and it was clear to see just from the files in the repo what the customizations were (instead of requiring a diff view). It was also easy to pin to monty versions and decide when to upgrade the version and make related changes. Is there anything that speaks against that workflow?

nielsleadholm replied:
Ah sorry, yes definitely worth mentioning it. I've added more discussion of alternative workflows, and tried to be less prescriptive in which one to follow (rather, just highlighting pros and cons).

vkakerbeck commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,104 @@
+
+If you are a member of the community working on a research idea, then we would recommend you also follow the above guidance under [Workflow for TBP Researchers: from Research Prototype to Implementation Project](#workflow-for-tbp-researchers-from-research-prototype-to-implementation-project). The main additional points to highlight are:
+- When creating the fork, you should set yourself as the owner (it will not be possible to set the Thousand Brains Project as the owner).
+- Please tag a Maintainer of the TBP when making atomic PRs into your feature branch. This will help get involvement from the TBP research team at an early stage of your work. Once you have tagged a Maintainer, we will triage and determine the most appropriate member of the team to review the PR.
```
This could just be a recommendation. I am not sure we need to require review of every atomic PR

vkakerbeck reviewed (commented):
Nice, thanks for writing out these guidelines. My main two high-level comments on content I was missing here:
- What about the monty-as-dependency approach (see more details in my comment on this)
- Will our cloud compute infrastructure work with this approach (i.e. can I run benchmark experiments from this feature fork the same way I do with my tbp.monty fork or do I need to make any updates to the cli tool?

nielsleadholm commented:
Thanks @vkakerbeck , good question about AWS infrastructure, I will check.

Re. the monty-as-dependency, I've written some thoughts, let me know what you think.

The other comments I have changed the text to address.

vkakerbeck commented:
Thanks for the updates! I left a comment on the monty-as-dependency question. The changes all look good to me :) Re. the AWS infrastructure, we don't need to add this info into this RFC but I think it is an important consideration for deciding our default setup.

tristanls commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,102 @@
+### Workflow for Contributing Code
+
+If you are a member of the community working on a research idea, then we would recommend you also follow the above guidance under [Workflow for TBP Researchers: from Research Prototype to Implementation Project](#workflow-for-tbp-researchers-from-research-prototype-to-implementation-project). The main additional points to highlight are:
+- When opening the fork, you should set yourself as the owner (it will not be possible to set the Thousand Brains Project as the owner).
```
@nielsleadholm, I was responding to Ramy's option 2.

nielsleadholm replied:
FYI I have added some clearer guidance for the option (1) that would work if someone does not have the ability to make multiple forks.

tristanls commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,104 @@
+    - However, researchers should use their own judgement to weigh the benefits of reviews on every atomic change against the benefit of quick iteration and experimentation.
+- During your work, ensure you regularly update your fork and its branches with any upstream changes to `thousandbrainsproject/tbp.monty`. This will reduce the chance that a longer RP-project is associated with significant merge conflicts when you later wish to merge it into the primary Monty code-base.
+- When a research prototype has been fully implemented and its merits established through evaluations and visualizations, the `main` branch is ready to form the basis of an Implementation Project.
+- The Implementation Project (IP) involves any necessary refactoring of the Research Prototype to ensure it adheres to the [guidelines of the platform](https://thousandbrainsproject.readme.io/docs/code-style-guide) before merging into `tbp.monty`. For more information, see [RFC 14 Conducting Research While Building a Stable Platform](https://github.com/thousandbrainsproject/tbp.monty/blob/main/rf ...
```
**suggestion (non-blocking)**: I think it'd be helpful to link directly to [the Implementation Project section](https://github.com/thousandbrainsproject/tbp.monty/blob/main/rfcs/0014_conducting_research_while_building_a_stable_platform.md#implementation-project) here.

tristanls commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,104 @@
+- During your work, ensure you regularly update your fork and its branches with any upstream changes to `thousandbrainsproject/tbp.monty`. This will reduce the chance that a longer RP-project is associated with significant merge conflicts when you later wish to merge it into the primary Monty code-base.
+- When a research prototype has been fully implemented and its merits established through evaluations and visualizations, the `main` branch is ready to form the basis of an Implementation Project.
+- The Implementation Project (IP) involves any necessary refactoring of the Research Prototype to ensure it adheres to the [guidelines of the platform](https://thousandbrainsproject.readme.io/docs/code-style-guide) before merging into `tbp.monty`. For more information, see [RFC 14 Conducting Research While Building a Stable Platform](https://github.com/thousandbrainsproject/tbp.monty/blob/main/rf ...
+- Note that the IP process will often involve a series of atomic PRs as well, rather than a single major PR. However, each one will be constrained to follow the platform guidelines.
```
**suggestion**: I think this might be better as a direct quote from RFC 14:

> The implementation project is a project because it may require major refactoring of the prototype or of Monty in order to integrate the new capability. It may be a single pull request, or it may be a long running effort as we reorganize all of Monty. The working prototype is helpful as it becomes a working specification of what needs to be implemented.
>
> This is the stage at which maintaining platform stability comes into consideration, not before.
>
> -- [RFC 14](https://github.com/thousandbrainsproject/tbp.monty/blob/main/rfcs/0014_conducting_research_while_building_a_stable_platform.md#implementation-project)

I think the summary here is overly restrictive by highlighting a series of atomic PRs.

Also, the note from RFC 14 may also be helpful:

> Please note that the decision as to when a prototype capability is stable, what constitutes a merge-worthy capability, and when to initiate the implementation project is at the discretion of the researcher.
>
> -- [RFC 14](https://github.com/thousandbrainsproject/tbp.monty/blob/main/rfcs/0014_conducting_research_while_building_a_stable_platform.md#rapid-prototyping-followed-by-platform-integration)

nielsleadholm commented:
Thanks for the feedback all, I made several updates which at a high level introduce more flexibility in the guidance. As part of this, I've added some diagrams to give a bit more clarity about the different workflows that can be used. It should be ready for review again. 

@hlee9212 re. your question about the IP process, I added a bit more detail on this, although I'm hesitant to add too much, as I feel like it will expand the scope significantly. It's also something that will come a lot down to @tristanls 's preferences, so is difficult for me to write out, and I think we can revisit this when we have an actual IP to go through. Hope that makes sense.

nielsleadholm commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,104 @@
+
+<img src="0000_code_guidance_for_researchers_and_community/create_new_RP_fork.png" alt="Create New Research Prototype Fork" width="500">
+
+Give your fork a name like `feat.description_of_prototype`:
```
Ok yeah I've added about the description, I don't recall either the specific guidance on RFC numbers for PRs, but agree it could be interesting. Will see what Ramy says.

nielsleadholm commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,104 @@
+- Researchers may also open PRs directly into `thousandbrainsproject/tbp.monty`, such as a refactor that is useful for them or other researchers, and which is not specific to their current feature project.
+- This will often be desirable if it constrains the complexity of a Research Prototype, and improves the quality of the platform. 
+- However, such PRs should be of general benefit if they are to be opened in this way. In other words, even if a Research Prototype is ultimately abandoned and archived, PRs directly into `tbp.monty` should still have been useful on their own merits.
+- When contemplating such PRs, it is important to consider when a change is suitable for a researcher to work on, vs. when it would benefit from an engineer's expertise. For example, major refactors, or features that are primarily about infrastructure tooling, are likely to be a poorer fit for a researcher's focus.
```
Ok thanks, the heading should already have TBP in it but I've added TBP in the text itself as well. I still think it is a useful thing to highlight and don't see it discouraging a community member.

ramyamounir reviewed (approved):
This looks great, thank you for adding in those changes!

tristanls commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,146 @@
+
+The workflow is shown in the below diagram:
+
+<img src="0000_code_guidance_for_researchers_and_community/main_workflow.png" alt="Workflow with a Feature Fork" width="450">
```
**praise**: Very nice graphic, summarizes the process well.

**issue**: This and the following graphics are missing the expected Researcher involvement in the Implementation Project. As such, I worry that it sets the wrong expectations.

>The general approach going forward is to enable rapid research prototypes without imposing platform stability requirements.
>
> 1. Researchers will create prototypes on their Monty forks and iterate rapidly there.
> 2. Once the prototypes demonstrate stable capability, **researchers will pair with engineers on an implementation project.**
> 3. The implementation project results in merging of the capability to Monty main.
>
> -- [RFC 14](https://github.com/thousandbrainsproject/tbp.monty/blob/main/rfcs/0014_conducting_research_while_building_a_stable_platform.md#rapid-prototyping-followed-by-platform-integration)

nielsleadholm replied:
Ok, the graphic is meant to show who is reviewing. I've added some description trying to make this clearer, let me know if you would like it changed further / still isn't clear.

tristanls replied:
Without the graphic itself changing, I believe the issue remains. 

**suggestion**:
<img width="583" height="237" alt="Screenshot 2025-08-28 at 11 40 52" src="https://github.com/user-attachments/assets/a4210af1-1f19-4387-a61e-e23c10bf4b84" />

nielsleadholm replied:
Ok thanks for the feedback, I've made these changes.

tristanls reviewed (commented):
I re-read the entire RFC.

One thing that strikes me about the RFC is that it is mainly about TBP Researchers. I struggle to find anything that a community member would find actionable. I think this is OK. 

**suggestion**: Make this an RFC for TBP Researchers only. The last two paragraphs mentioning the community could be folded into the existing documentation. I think this would be clearer for readers. The curious community members could still read through to see how we organize our research and extract what they want from it. But, more importantly, community members wouldn't feel that they have to do anything suggested here to contribute to research. Instead, they'd follow the normal GitHub norms (as much as there are any 🙂)

nielsleadholm commented on rfcs/0000_code_guidance_for_researchers_and_community.md:
```diff
@@ -0,0 +1,146 @@
+- When you open a PR into your fork's `main` branch, it is generally recommended that you tag at least one other TBP researcher for review.
+    - The aim is that RP work should be regularly reviewed and discussed with other researchers; getting reviews early, particularly for complex or conceptually important changes, reduces the chance that significant effort will be spent on a change that might be in a suboptimal direction.
+    - However, researchers should use their own judgement to weigh the benefits of reviews on every atomic change against the benefit of quick iteration and experimentation.
+- During your work, ensure you regularly update your fork and its branches with any upstream changes to `thousandbrainsproject/tbp.monty`. This will reduce the chance that a longer RP-project is associated with significant merge conflicts when you later wish to merge it into the primary Monty code-base.
```
Hmmm yeah I would maybe suggest a middle-of-the-road approach. The reality is most upstream changes are fine, but if left for long (and RPs do sometimes drag on), they can accumulate to a bit of a nightmare. I might soften the language to suggest that it is advised to regularly pull from upstream, but if this ever introduces breaking changes, then rather than try to resolve them the, the recommendation is to revert to the older commit and deal with it later during the IP.

tristanls replied:
@nielsleadholm, that sounds good 👍

tristanls commented:
> but my hope is that it will encourage others to feel like they can contribute to the research-side of things, i.e. that we don't assume that only TBP researchers will be doing this.

@nielsleadholm, I honestly don't think that this RFC is a good mechanism to encourage outside researchers. I think there ought to be better ways of doing it. Perhaps we could get an outside researcher to read this RFC and see if they are encouraged? I guess that this RFC will have the opposite effect by highlighting a lot of complexity that outside researchers don't need to deal with or care about.

vkakerbeck reviewed (commented):
Nice, thanks for the updates! Looks great :) I left a few more small comments.
Did you have a chance to check whether our AWS setup works out of the box with the RP forks?
